### PR TITLE
Associate heuristic rules with file extensions

### LINKED ISSUE #2042: Errors when searching for the correct heuristic rule to use
I found an issue with the heuristics and the way it looks for the right heuristic to use.
In the [Heuristic strategy](https://github.com/github/linguist/blob/master/lib/linguist/heuristics.rb#L54-L57), these lines checks if a certain heuristic matches the candidate languages:
```ruby
# Internal: Check if this heuristic matches the candidate languages.
def matches?(candidates)
  candidates.any? && candidates.all? { |l| @languages.include?(l.name) }
end
```
It basically says that all candidate languages must be in the list of language this heuristic is trying to distinguish.

0. So if one of the candidate languages is not in the list, the heuristic will not be used.
0. It also means that, if the heuristic was written to distinguish more languages than we have candidates, it will still be used.

At the time we write the heuristic we generally make sure that the first case doesn't happen. For example when we wrote the heuristic for AsciiDoc and AGS Script (`.asc files`) they were the only two languages using the `.asc` extension. But since, Public Key has been added with the `.asc` extension. Thus the heuristic for AsciiDoc and AGS Script is not used anymore when trying to distinguish `.asc` files. This is why the issue #1638 is still opened.
We also wrote a heuristic rule for Prolog and IDL on `.pro` files. But now this file extension is also defined for [QMake](https://github.com/github/linguist/blob/master/lib/linguist/languages.yml#L2417), explaining why [so many `.pro` files are still not correctly detected](https://github.com/search?l=idl&q=extension%3Apro+is&ref=searchresults&type=Code&utf8=%E2%9C%93).

The second point explains the issue with [these Perl files](https://github.com/rakudo/rakudo/search?l=prolog) (reported by @hoelzro in #1680). The `.pm` extension causes a conflict between Perl 5 and 6. Then Linguist uses the heuristic for Perl, Perl6 and Prolog because it includes all the candidate languages it's trying to distinguish. Unfortunately, ([because our heuristic doesn't check for comments](https://github.com/rakudo/rakudo/blob/421176c14605dcca7d339aec2d61aaee329b9f6d/src/core/io_operators.pm#L162)) Prolog is chosen as the correct language. Even if it wasn't a candidate for `.pm` files...

I'm not really sure on how this should be fixed. I also don't think I have enough experience with Ruby to write a test that detects these cases happening (I tried quickly but hit some specificities to Ruby's OOP).

/cc @bkeepers @arfon

### LINKED ISSUE #2149: Some .pm categorised as Perl6, some as Perl
Hey,

We have a collection of `.pm` files, all of which are Perl 5; however about 10% are being classed as Perl 6 incorrectly:

* https://github.com/duckduckgo/zeroclickinfo-goodies/search?l=perl6
* https://github.com/duckduckgo/zeroclickinfo-goodies/search?l=perl

I've had a quick look and I can't see what the particular reason is. A specific example is:
Erroneously Perl 6:
* https://github.com/duckduckgo/zeroclickinfo-goodies/blob/f352c6c9ae215c68607c38acaa9b89402fe736a3/lib/DDG/Goodie/DateMath.pm
* https://github.com/duckduckgo/zeroclickinfo-goodies/blob/f352c6c9ae215c68607c38acaa9b89402fe736a3/lib/DDG/Goodie/PrivateNetwork.pm

Correctly Perl 5:
* https://github.com/duckduckgo/zeroclickinfo-goodies/blob/f352c6c9ae215c68607c38acaa9b89402fe736a3/lib/DDG/Goodie/BashPrimaryExpressions.pm
* https://github.com/duckduckgo/zeroclickinfo-goodies/blob/f352c6c9ae215c68607c38acaa9b89402fe736a3/lib/DDG/Goodie/TimezoneConverter.pm

cc @jagtalon @mrchrisw

### PR DESCRIPTION
**Pull request by pchaigno**
This pull request associates one or several file extensions to each heuristic rule instead of languages as discussed in #2359.
Fixed #2042.

be122ca also improves the heuristic rule for Perl6 and fixes #2149.

### CODE REVIEW

File: lib/linguist/heuristics.rb [hunk 1]
```diff
@@ -13,11 +13,14 @@ class Heuristics
...
     # Returns an Array of languages, or empty if none matched or were inconclusive.
-    def self.call(blob, languages)
+    def self.call(blob, candidates)
       data = blob.data
 
       @heuristics.each do |heuristic|
-        return Array(heuristic.call(data)) if heuristic.matches?(languages)
+        if heuristic.matches?(blob.name)
+          languages = Array(heuristic.call(data))
+          return languages if languages.any? || languages.all? { |l| candidates.include?(l) }
```
Comment (pchaigno) [inline lib/linguist/heuristics.rb]: `languages.any? || languages.all? { |l| candidates.include?(l) }` should always be true.
Maybe we should throw a warning when it's not...?

**Comment by Languagelearner1111** [comment]
I probably did the pull request by accident since I don't actually know
what a pull request is. Coding noon here.

**Comment by larsbrinkhoff** [comment]
I think this is an improvement.  Good job!

**Comment by bkeepers** [comment]
:+1: I like this approach.

@arfon?

**Comment by yyx990803** [comment]
+1

Most importantly, this allows excluding non-listed languages from a given extension.

**Comment by arfon** [comment]
:+1: thanks @pchaigno

**Pull request merged**